### ISSUE TITLE
flash attn triton version may have bf16 overflow.

### ISSUE THREAD

**Issue by shiqingzhangCSU** [Dao-AILab/flash-attention#442]
When we tested the random mask, we found that if the value in the mask is large, the flash attn implementation of triton is different from the standard attention. Maybe this code exceeds the range indicated by bf16 during calculation? Can you help analyze it?

my code show as below：
```

import math
import numpy as np
import torch
from flash_attn.flash_attn_interface import flash_attn_varlen_func
from flash_attn.flash_attn_triton import flash_attn_func

def self_attention_reference(q, k, v, attention_mask):
    bsz, q_len, num_heads, head_dim = q.shape
    q = q.transpose(1, 2)
    k = k.transpose(1, 2)
    v = v.transpose(1, 2)

    attn_weights = torch.matmul(q, k.transpose(2, 3)) / math.sqrt(head_dim)

    if attention_mask is not None:
        attn_weights = attn_weights + attention_mask
        attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))
    attn_weights_save = attn_weights
    attn_weights = torch.nn.functional.softmax(attn_weights, dim=-1)
    attn_weights32 = torch.nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(v.dtype)
    attn_output = torch.matmul(attn_weights, v)
    return attn_output


def flash_attention_triton(q, k, v, attention_mask):
    attn_output = flash_attn_func(q, k, v, attention_mask, False)
    return attn_output

def _fill_with_neg_inf(t):
    """FP16-compatible function that fills a tensor with -inf."""
    return t.float().fill_(float("-inf")).type_as(t)


# Copied from transformers.models.bart.modeling_bart._expand_mask
def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len = None):
    """
    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
    """
    bsz, src_len = mask.size()
    tgt_len = tgt_len if tgt_len is not None else src_len

    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)

    inverted_mask = 1.0 - expanded_mask

    return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)


def _get_interleave(n):
        def _get_interleave_power_of_2(n):
            start = (2 ** (-2 ** -(math.log2(n) - 3)))
            ratio = start
            return [start * ratio ** i for i in range(n)]

        if math.log2(n).is_integer():
            return _get_interleave_power_of_2(n)
        else:
            closest_power_of_2 = 2 ** math.floor(math.log2(n))
            return _get_interleave_power_of_2(closest_power_of_2) + \
                _get_interleave(2 * closest_power_of_2)[0::2][:n - closest_power_of_2]


def _fill_with_neg_inf(t):
    """FP16-compatible function that fills a tensor with -inf."""
    return t.float().fill_(float("-inf")).type_as(t)


def _gen_parallel_alibi_mask(n_head, max_pos):
    tp_world_size = 1
    tp_rank = 0

    slopes = torch.Tensor(_get_interleave(n_head))
    alibi = slopes.unsqueeze(1).unsqueeze(1) * torch.arange(max_pos).unsqueeze(0).unsqueeze(0).expand(
        n_head, -1, -1)
    alibi = alibi.view(n_head, 1, max_pos)
    alibi_mask = torch.triu(
        _fill_with_neg_inf(torch.zeros([max_pos, max_pos])), 1
    )
    alibi_mask = alibi_mask.unsqueeze(0) + alibi

    assert n_head % tp_world_size == 0, \
        f'n_head is not dividable by tp_world_size. n_head: {n_head}, tp_world_size: {tp_world_size}'
    begin_idx = tp_rank * (n_head // tp_world_size)
    end_idx = (tp_rank + 1) * (n_head // tp_world_size)
    alibi_mask = alibi_mask[begin_idx:end_idx]

    return alibi_mask


if __name__ == '__main__':
    size = 512
    bsz, q_len, num_heads, head_dim = 1, size, 40, 64
    qkv_shape = (bsz, q_len, num_heads, head_dim)
    device = "cuda"
    dtype = torch.bfloat16

    q = torch.rand(qkv_shape, device=device, dtype=dtype)
    k = torch.rand(qkv_shape, device=device, dtype=dtype)
    v = torch.rand(qkv_shape, device=device, dtype=dtype)
    attention_mask = 200*torch.rand(1, num_heads, q_len, q_len, device=device, dtype=dtype)

    # it work
    # attention_mask = *torch.rand(1, num_heads, q_len, q_len, device=device, dtype=dtype)  
    
    # attention_mask = torch.zeros_like(attention_mask)
    mask2 = attention_mask + torch.triu(_fill_with_neg_inf(torch.zeros([size, size], dtype=dtype)), 1).unsqueeze(0).to("cuda")

    result_reference = self_attention_reference(q, k, v, mask2)
    result_flash = flash_attention_triton(q, k, v, mask2)
    result_flash = result_flash.transpose(2, 1)

    # result_af = torch.randn_like(q)
    # norm_factor = 1/math.sqrt(128)
    # result_af = attention_forward(q, k, v, result_af, norm_factor, True, mask2)


    result_reference_numpy = result_reference.float().cpu().detach().numpy()
    result_flash_numpy = result_flash.float().cpu().detach().numpy()

    if not np.allclose(result_reference_numpy, result_flash_numpy, 1e-4, 1e-4):
        diff = np.abs(result_reference_numpy - result_flash_numpy)
        print("model output diff larger than 1e-4")
        print("max diff:  %f" % np.max(diff))
        index = np.argmax(diff)
        print("index: %d     value: %f  %f" % (index, result_reference_numpy.flatten()[index], result_flash_numpy.flatten()[index]))
        print("")
        for i in range(20):
            idx = index - 10 + i 
            print("index: %d     value: %f  %f" % (idx, result_reference_numpy.flatten()[idx], result_flash_numpy.flatten()[idx]))

```

**Comment by janelu9** [comment]
As I know, only `triton==2.0.0.dev20221202` was supported.
